packages/core/schematics/test/invalid_two_way_bindings_spec.ts

```diff
@@ -0,0 +1,341 @@
+      `</span>`,
+      `</div>`,
+    ].join('\n'));
+  });
+
+  it('should migrate multiple two-way bindings on the same element', async () => {
+    writeFile('/comp.ts', `
+      import {Component} from '@angular/core';
+
+      @Component({
+        template: \`<input [(foo)]="a && b" bar="123" [(baz)]="!!a"/>\`
+      })
+      class Comp {}
+    `);
+
+    await runMigration();
+    const content = tree.readContent('/comp.ts');
+    expect(content).toContain(
+        'template: \`<input [foo]="a && b" (fooChange)="a && (b = $event)" ' +
+        'bar="123" [baz]="!!a" (bazChange)="a = $event"/>\`');
+  });
+
+  it('should not stop the migration if a file cannot be read', async () => {
+    writeFile('/comp.ts', `
+      import {Component} from '@angular/core';
+
+      @Component({
+        templateUrl: './does-not-exist.html'
+      })
+      class BrokenComp {}
+    `);
+
+    writeFile('/other-comp.ts', `
+      import {Component} from '@angular/core';
+
+      @Component({
+        templateUrl: './comp.html'
+      })
+      class Comp {}
+    `);
+
+    writeFile('/comp.html', '<input [(ngModel)]="a || b"/>');
+
+    await runMigration();
+    const content = tree.readContent('/comp.html');
+
+    expect(content).toBe('<input [ngModel]="a || b" (ngModelChange)="a || (b = $event)"/>');
+  });
+
+  it('should migrate a component that is not at the top level', async () => {
+    writeFile('/comp.ts', `
+      import {Component} from '@angular/core';
+
+      function foo() {
+        @Component({
+          template: \`<input [(ngModel)]="a || b"/>\`
+        })
+        class Comp {}
+      }
+    `);
+
+    await runMigration();
+    const content = tree.readContent('/comp.ts');
+
+    expect(content).toContain(
+        'template: `<input [ngModel]="a || b" (ngModelChange)="a || (b = $event)"/>`');
+  });
+
+  it('should preserve a valid expression', async () => {
+    writeFile('/comp.ts', `
+      import {Component} from '@angular/core';
+
+      @Component({
+        template: \`<input [(ngModel)]="a.b.c"/>\`
+      })
+      class Comp {}
+    `);
+
+    await runMigration();
+    const content = tree.readContent('/comp.ts');
+    expect(content).toContain('template: `<input [(ngModel)]="a.b.c"/>`');
+  });
+
+  it('should not migrate an invalid expression if an event listener for the same event exists',
+     async () => {
+       writeFile('/comp.ts', `
+        import {Component} from '@angular/core';
+
+        @Component({
+          template: \`<input [(ngModel)]="a || b" (ngModelChange)="foo($event)"/>\`
+        })
+        class Comp {}
+      `);
+
+       await runMigration();
+       const content = tree.readContent('/comp.ts');
+       expect(content).toContain(
+           'template: `<input [(ngModel)]="a || b" (ngModelChange)="foo($event)"/>`');
+     });
+
+  it('should not migrate an invalid expression if a property binding for the same event exists',
```

Comment:
Nit, but honestly neither is making me feel good 😄 

```suggestion
  it('should not migrate an invalid expression if a one-way property binding for the same binding exists',
```